### DISCUSSION TITLE
Strongly-typed AzureCredentials in httpclient.Options

### DISCUSSION THREAD

**Discussion by kostrse** [grafana/grafana#37044] [Data Sources]
PR #36932 adds experimental support for Azure authentication directly into `httpclientprovider`  to make it possible for any datasource to activate it by providing proper settings via `httpclient.Options`.

Currently, since `httpclient.Options` does not provide any strongly-typed configuration for Azure, the `CustomOptions` key-value bag is used to pass the needed settings.

### AzureCredentials in httpclient.Options

The proposal is to add strongly-typed configuration for Azure credentials into the `httpclient.Options`  structure, similar to `BasicAuth`, `SigV4` etc.

`httpclient.Options` is part of Grafana Plugin SDK.

```go
package httpclient
...
type Options struct {
    ...
    AzureCredentials azcredentials.AzureCredentials
    // or
    Azure: &AzureSettings{
        Credentials azcredentials.AzureCredentials
    }
    ...
}
```

### Dependencies between packages

Currently, there are two Azure packages which are candidates to be moved out of Grafana codebase into a separate repo [grafana-azure-sdk-go](https://github.com/grafana/grafana-azure-sdk-go):

* [azcredentials](https://github.com/grafana/grafana/tree/3aac4ab3915fc62ae505488c758c1d8de2a4ee63/pkg/tsdb/azuremonitor/azcredentials) - contains only definitions of `AzureCredentials` type and builders/constructors. Doesn't have any implementation and doesn't have any dependencies. Intended only for interfaces / models.

* [aztokenprovider](https://github.com/grafana/grafana/tree/3aac4ab3915fc62ae505488c758c1d8de2a4ee63/pkg/tsdb/azuremonitor/aztokenprovider) - contains actual implementation of token provider, middleware. Has dependencies on Grafana, Grafana SDK and other libraries like Azure SDK for Go etc.

To reference `AzureCredentials` in `httpclient.Options`,  Grafana Plugin SDK (`grafana-plugin-sdk-go`) would need to have dependency on  the`github.com/grafana/grafana-azure-sdk-go/pkg/azcredentials` package which contains definition of `AzureCredentials`. 

The `aztokenprovider` would have dependency on `azcredentials` and  `grafana-plugin-sdk-go`:

`azcredentials` :arrow_left: `grafana-plugin-sdk-go` :arrow_left: `aztokenprovider` :arrow_left: Grafana / plugins

Alternative option is to eliminate `azcredentials`, add definition of `AzureCredentials` directly into Grafana Plugin SDK:

`grafana-plugin-sdk-go` :arrow_left: `aztokenprovider` :arrow_left: Grafana / plugins

### Related work

_TBD_

**Comment by pkolyvas** [comment]
@kostrse Thanks for the PR - it's been merged and should be available in `Grafana 8.1.0-beta2` !
